wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -317,41 +329,66 @@ public boolean hasRuntime(String workspaceId) {
-                    if (runtimes.putIfAbsent(id.getWorkspaceId(),
-                                             new RuntimeState(validate(infra.getRuntime(id)), RUNNING)) != null) {
-                        // should not happen, violation of SPI contract
-                        LOG.error("More than 1 runtime of workspace found. " +
-                                  "Runtime identity of duplicate is '{}'. Skipping duplicate.", id);
-                    }
+                for (RuntimeIdentity identity : infra.getIdentities()) {
+                    recoverOne(infra, identity);
                 }
             } catch (UnsupportedOperationException x) {
                 LOG.warn("Not recoverable infrastructure: '{}'", infra.getName());
-            } catch (InfrastructureException x) {
+            } catch (ServerException | InfrastructureException x) {
                 LOG.error("An error occurred while attempted to recover runtimes using infrastructure '{}'. Reason: '{}'",
-                          infra.getName(),
-                          x.getMessage());
+                          infra.getName(), x.getMessage());
             }
         }
     }
 
-    private void subscribeCleanupOnAbnormalRuntimeStopEvent() {
-        eventService.subscribe(new CleanupRuntimeOnAbnormalRuntimeStop());
+    private void recoverOne(RuntimeInfrastructure infra, RuntimeIdentity identity) throws ServerException {
+        Workspace workspace;
+        try {
+            workspace = workspaceDao.get(identity.getWorkspaceId());
+        } catch (NotFoundException x) {
+            LOG.error("Workspace configuration is missing for the runtime '{}:{}'. Runtime won't be recovered",
+                      identity.getWorkspaceId(),
+                      identity.getEnvName());
+            return;
+        }
+
+        Environment environment = workspace.getConfig().getEnvironments().get(identity.getEnvName());
+        if (environment == null) {
+            LOG.error("Environment configuration is missing for the runtime '{}:{}'. Runtime won't be recovered",
+                      identity.getWorkspaceId(),
+                      identity.getEnvName());
+            return;
+        }
+
+        InternalRuntime runtime;
+        try {
+            runtime = infra.prepare(identity, environment).getRuntime();
+        } catch (InfrastructureException | ValidationException x) {
+            LOG.error("Couldn't recover runtime '{}:{}'. Error: {}",
+                      identity.getWorkspaceId(),
+                      identity.getEnvName(),
+                      x.getMessage());
+            return;
+        }
+
+        RuntimeState prev = runtimes.putIfAbsent(identity.getWorkspaceId(), new RuntimeState(runtime, RUNNING));
+        if (prev == null) {
+            LOG.info("Successfully recovered workspace runtime '{}'", identity.getWorkspaceId(), identity.getEnvName());
+        } else {
+            LOG.error("More than 1 runtime with id '{}:{}' found. " +
```

Comment:
I was thinking about the following use-case:
1. someone starts che with test configuration/bindings
2. someone sees the warning that workspace is not recovered because of the collision 
3. someone fixes configuration and restarts che
4. someone gets workspace recovered =)

Obviously it shouldn't happen in normal conditions, but if it happens let admin decide what to do with running containers. Maybe we will extract this code in the strategy so the behaviour can be configurable, but i won't do it here.